students/semko_krzysztof/lesson_03_functions/input_validation.py

```diff
@@ -0,0 +1,40 @@
+"""
+Add try and except statements to the previous project to detect
+whether the user types in a noninteger string. Normally, the int()
+function will raise a ValueError error if it is passed a noninteger
+string, as in int('puppy'). In the except clause, print a message
+to the user saying they must enter an integer.
+"""
+
+
+def collatz(number):
+    if number < 0:
+        print("This is not a positive integer.")
+        return -1
+    elif number % 2 == 0:
+        result = number // 2
+        print(str(result))
+        return result
+    elif number % 2 == 1:
+        result = 3 * number + 1
+        print(str(result))
+        return result
+    else:
+        print("This is not a positive integer.")
+        return -1
+
+
+value = -1
+while value != 1:
+    try:
+        if value == -1:
+            print("Please input a number:")
+            value = float(input()) // 1
```

Comment:
1. Jeżeli wczytujemy input z klawiatury tylko raz, to należy pomyśleć czy nie dałoby się tego przenieść przed pętlę, Jeżeli chcesz wczytywać do skutku aż user wprowadzi poprawna wartość to zrobiłbym jedną pętlę na wczytanie danej a drugą do obliczania collatz sequence. Wtedy poszczególne kroku programu będą bardziej rozdzielone i zyskamy na czytelności. Zalecam porawkę.
2. Zamiast `float(input()) // 1` prościej zrobić `value = int(input())` i traktować floaty jako niepoprawne dane, chyba że jest coś o czym nie pomyślałem?

Reply:
1. Z pełną premedytacją skorzystałem tutaj z jednej pętli. Chciałem uniknąć dwukrotnego sprawdzania poprawności wprowadzonej wartości - uznałem, że taki warunek i tak powinien znaleźć się w collatz(), a skoro tak to można oprzeć pętlę na zwracanej przez funkcję wartości.
2.  Racja. Ale w samym the_collatz_sequence.py float() powinien wg. mnie pozostać (przykładowy input: 0.23)